app/policies/service_provider_mfa_policy.rb

```diff
@@ -0,0 +1,80 @@
+class ServiceProviderMfaPolicy
+  AAL3_METHODS = %w[webauthn piv_cac].freeze
+
+  attr_reader :mfa_context, :auth_method, :service_provider
+
+  def initialize(
+    user:,
+    service_provider:,
+    auth_method:,
+    aal_level_requested:,
+    piv_cac_requested:
+  )
+    @user = user
+    @mfa_context = MfaContext.new(user)
+    @auth_method = auth_method
+    @service_provider = service_provider
+    @aal_level_requested = aal_level_requested
+    @piv_cac_requested = piv_cac_requested
+  end
+
+  def user_needs_sp_auth_method_verification?
+    # If the user needs to setup a new MFA method, return false so they go to
+    # setup instead of verification
+    return false if user_needs_sp_auth_method_setup?
+
+    if piv_cac_required?
+      auth_method.to_s != 'piv_cac'
+    elsif aal3_required?
+      !AAL3_METHODS.include?(auth_method.to_s)
+    else
+      false
+    end
+  end
+
+  def user_needs_sp_auth_method_setup?
+    return true if piv_cac_required? && !piv_cac_enabled?
+    return true if aal3_required? && !aal3_enabled?
+    false
+  end
+
+  def auth_method_confirms_to_sp_request?
+    !user_needs_sp_auth_method_setup? && !user_needs_sp_auth_method_verification?
+  end
+
+  def aal3_required?
+    aal3_requested? || service_provider&.aal == 3
+  end
+
+  def piv_cac_required?
+    piv_cac_requested?
+  end
+
+  def multiple_required_methods_enabled?
```

Comment:
I'm getting hung up on the name of this method. If aal3 is not required, this returns true, even though the user may only have one mfa enabled. Maybe this should be called `allow_user_to_switch_method?` or something?

Reply:
I like that, let's do it